community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java

```diff
@@ -0,0 +1,134 @@
+
+import org.neo4j.concurrent.AsyncEvents;
+import org.neo4j.function.Consumer;
+import org.neo4j.kernel.impl.util.CappedOperation;
+import org.neo4j.kernel.impl.util.Listener;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import static org.neo4j.kernel.impl.util.CappedOperation.differentItemClasses;
+import static org.neo4j.kernel.impl.util.CappedOperation.time;
+
+/**
+ * Wraps a {@link LogProvider}, making its {@link Log logs} asynchronous by letting their log statements be
+ * treated as events fed into a {@link AsyncEvents} to be processed by a dedicated logging thread instead
+ * of in the thread making the log statement call.
+ */
+class AsyncLogProvider implements LogProvider, Closeable, Runnable
+{
+    private final AsyncEvents<AsyncLogEvent> events;
+    private final LogProvider actual;
+
+    AsyncLogProvider( LogProvider actual )
+    {
+        this( actual, new Listener<Throwable>()
+        {
+            @Override
+            public void receive( Throwable error )
+            {
+                error.printStackTrace();
+            }
+        } );
+    }
+
+    AsyncLogProvider( LogProvider actual, final Listener<Throwable> errorHandler )
+    {
+        this.events = new AsyncEvents<>( new Consumer<AsyncLogEvent>()
+        {
+            private final CappedOperation<Throwable> errorNotification = new CappedOperation<Throwable>(
+                    differentItemClasses(), time( 1, SECONDS ) )
+            {
+                @Override
+                protected void triggered( Throwable error )
+                {
+                    errorHandler.receive( error );
+                }
+            };
+
+            @Override
+            public void accept( AsyncLogEvent logEvent )
+            {
+                try
+                {
+                    logEvent.process();
+                }
+                catch ( Throwable t )
+                {
+                    // We catch Throwable here otherwise there's a high chance an error, like OutOfMemory
+                    // or similar will halt the logging from that point and onwards, since all logging
+                    // is processed by a dedicated thread.
+                    //
+                    // What's most important is that logging gets going, but we also can be nice and...
+                    // well... not log, since apparently there was a problem doing that right now, but do something.
+                    // That something will be to print stack trace of the exception, capped by different
+                    // exception types and time interval for good measure.
+                    errorNotification.event( t );
+                }
+            }
+        } );
+        this.actual = actual;
+    }
+
+    @Override
+    public Log getLog( @SuppressWarnings( "rawtypes" ) Class loggingClass )
+    {
+        return new AsyncLog( events, actual.getLog( loggingClass ) );
```

Comment:
It is probably a better idea to cache here `AsyncLog`s rather than in `AbstractLogProvider`.